### ISSUE TITLE
Refereced objects that use types.union died when applying snapshot

### ISSUE THREAD

**Issue by abruzzihraig** [mobxjs/mobx-state-tree#134]
Hi guys,
I have to say this is an amazing project. It always gives me all kinds of surprises when I use it to build a project with state management and observable features. I really appreciate your talent work.

But today I got a weird issue and it took me a few hours to dig until now. I think the document is not perfect by now, and I still don't really understand parts of them. So I want to describe the confusing issue I met to you, then you may help me to know if it is indeed an issue from the mobx-state-tree or just a wrong way I wrote. Thanks.

The scene has a little bit complicated so I try my best to simplify it for you.

##### Stage 1
At the beginning, there was a simple store that held two properties like below:

```
const BuildingStore = types.model('BuildingStore', {
    objects: types.array(types.union(Text, Post)),
    selection: types.reference(types.union(Post, Text))
}, {
    setSelection(selection: IUIObject) {
        if (this.selection === selection) return;

        if (this.selection && isAlive(this.selection)) {
            if (this.selection.setEditablility) {
                this.selection.setEditablility(false);
            }
        }

        this.selection = selection;
})
```

As you can see, I used `types.union` there in both of `types.array` and `types.reference`. I think it makes sense although I cannot find any example for how to use `types.union` in `types.reference`.

During this stage, I got an error when I was trying to `applySnapshot` to `BuildingStore`. The error message showed it's an ambiguous snapshot. So I just put a dispatcher function for the property `objects`:

##### Stage 2

```
const BuildingStore = types.model('BuildingStore', {
    objects: types.array(types.union((snapshot: any) => {
        if (snapshot.type === 'Post') return Post;
        if (snapshot.type === 'Text') return Text;
    }, Post, Text)),
    selection: types.reference(types.union(Post, Text))
})
```

At this time, the ambiguous error disappeared. But when I did something like undo to `applySnapshot` to `BuildingStore` again, I found I cannot get the computed properties from the `selection` anymore. At the mean time, if I do an operation that triggers the method `setSelection`, I will get a new error like below:

```
utils.js:14 Uncaught Error: [mobx-state-tree] Failed to assign a value to a reference; the value should already be part of the same model tree
    at Object.invariant (utils.js:14)
    at Reference.setNewValue (reference.js:64)
    at BuildingStore.set (reference-property.js:40)
    at BuildingStore.setSelection (buildingStore.ts:22)
    at executeAction (mobx.js:663)
    at BuildingStore.res (mobx.js:654)
    at BuildingStore.actionInvoker (action.js:38)
    at BuildingStore.setSelection (eval at createNamedFunction (utils.js:110), <anonymous>:3:43)
    at runRawAction (action.js:5)
    at runMiddleWares (action.js:21)
```

So I was confusing about when the value out of the model tree. Then I found something more in the following stage:

##### Stage 3

```
export const Post = types.model('Post', {
    id: types.identifier(),
    type: types.string,
    content: types.string,
    isEditable: types.boolean,
    fontSize: types.number,
    fontSizeUnit: types.union(types.literal('em'), types.literal('px'), types.literal('pt'), types.literal('')),
    lineHeight: types.number,
    lineHeightUnit: types.union(types.literal('em'), types.literal('px'), types.literal('pt'), types.literal('')),
    color: types.string,
    textAlign: types.string,
    get isActive(): boolean {
        if (!hasParent(this)) return false; // hasParent reterns false
        return buildingStore.selection === this;
    },
    get parsedContent(): string {
        return marked(this.content);
    }
}
```

When I apply a snapshot to `BuildingStore`, the getter above `isActive` will run again, but this time the function `hasParent` returns `false`. That means `this` post instance has already left from the model tree.

So I tried to learn more from `types.union` then I found the code:
https://github.com/mobxjs/mobx-state-tree/blob/master/src/types/utility-types/union.ts#L36
I don't really understand those source code, but just guess if there is something wrong during the step creating a new instance in `types.union`.

Otherwise, I've tried to add a dispatcher to the reference property too:

```
const BuildingStore = types.model('BuildingStore', {
    objects: types.array(types.union((snapshot: any) => {
        if (snapshot.type === 'Post') return Post;
        if (snapshot.type === 'Text') return Text;
    }, Post, Text)),
    selection: types.reference(types.union((snapshot: any) => {
        if (snapshot.type === 'Post') return Post;
        if (snapshot.type === 'Text') return Text;
    }, Post, Text))
}
```

But there's nothing changed. I even don't know whether I need to add a dispatcher to it.

I also tried something like below:

```
types.union(types.reference(Text), types.reference(Post))
```

It doesn't have a syntax error, and the document also doesn't say which is the right way to combine them, but I still got the error in runtime:

```
Uncaught TypeError: type.is is not a function
    at http://0.0.0.0:8100/bundle.js:16699:59
    at Array.some (native)
    at Union.validate (http://0.0.0.0:8100/bundle.js:16699:24)
    at ValueProperty.validate (http://0.0.0.0:8100/bundle.js:28546:26)
    at http://0.0.0.0:8100/bundle.js:11535:120
    at Array.map (native)
    at ObjectType.isValidSnapshot (http://0.0.0.0:8100/bundle.js:11535:73)
    at ObjectType.ComplexType.validate (http://0.0.0.0:8100/bundle.js:10364:21)
    at Object.typecheck (http://0.0.0.0:8100/bundle.js:1604:23)
    at ObjectType.ComplexType.create (http://0.0.0.0:8100/bundle.js:10334:24)
```

That's all. I really hope you can help me on the error that I got in Stage 2. Thanks!

**Comment by abruzzihraig** [comment]
A further test by myself found that the bug it has nothing to the dispatcher function. Once you apply a snapshot to a store that defined by `types.array(types.union(A, B))`, you will get died objects in the new array. But `types.array(A)` or `types.array(B)` work fine.

**Comment by abruzzihraig** [comment]
After a deep debugging, I found when applying a snapshot to a property defined by `types.array(A)`, it will call the method `MSTAdministration.prototype.reconcileChildren`. I think this method will replace the old value with a new value without creating a new instance.

But when the property is changed to `types.array(types.union(A, B))`, it won't call the method `reconcileChildren`. And the method `ComplexType.proptotype.create` will be called instead. Then I got objects without parents. So I guess maybe the problem comes from there.

**Comment by abruzzihraig** [comment]
@mattiamanzati Hi, I just wonder if there's any progress on this issue? If you need, I could create a minimum demo for presenting the bug.

**Comment by abruzzihraig** [comment]
@mattiamanzati Here's the minimum demo for presenting the bug, and it can be easily run by Node.js:

```
const uuid = require('uuid');
const mst = require('mobx-state-tree');
const { types, hasParent, tryResolve } = mst;

const Cat = types.model('Cat', {
    id: types.identifier(),
    type: 'Cat',
    name: types.string,
    get path() {
        return getPath(this);
    }
});

const Dog = types.model('Dog', {
    id: types.identifier(),
    type: 'Dog',
    name: types.string
});

const Store = types.model('Store', {
    animals: types.array(types.union((snapshot) => {
        if (snapshot.type === 'Cat') return Cat;
        if (snapshot.type === 'Dog') return Dog;
    }, Cat, Dog)),
    cats: types.array(Cat),
    selection: types.reference(types.union(Cat, Dog))
}, {
    setSelection(selection) {
        if (this.selection === selection) return;

        console.info(`whether the selection ${selection.name} has parent: `, hasParent(selection));

        this.selection = selection;
    },
    setObject(obj) {
        this.animals.push(obj);
        return this.animals[this.animals.length-1];
    }
});

const store = Store.create({
    animals: [{
        id: uuid.v1(),
        type: 'Cat',
        name: 'Kitty'
    }, {
        id: uuid.v1(),
        type: 'Dog',
        name: 'Meat'
    }],
    cats: [{
        id: uuid.v1(),
        type: 'Cat',
        name: 'Jazz'
    }, {
        id: uuid.v1(),
        type: 'Cat',
        name: 'Punk'
    }],
    selection: null
});


const initializedAnimal = tryResolve(store, './animals/0');

const initializedCat = tryResolve(store, './cats/0');

const additionalAnimal = store.setObject(Cat.create({
    id: uuid.v1(),
    type: 'Cat',
    name: 'Ruby'
}));

console.info('initializedAnimal: ', initializedAnimal.toJSON());
console.info('initializedCat: ', initializedCat.toJSON());
console.info('additionalAnimal: ', additionalAnimal.toJSON());


store.setSelection(additionalAnimal);

console.info('set selection on additionalAnimal successfully.');

store.setSelection(initializedCat);

console.info('set selection on initializedCat successfully.');

store.setSelection(initializedAnimal); // It will break with the error 'Failed to assign a value to a reference; the value should already be part of the same model tree'

console.info('set selection on initializedAnimal successfully.');
```

And the detailed logs output like below:

```
initializedAnimal:  {
  id: 'e13759c0-4111-11e7-a34f-2b9b5e15da6e',
  type: 'Cat',
  name: 'Kitty' 
}
initializedCat:  { 
  id: 'e13759c2-4111-11e7-a34f-2b9b5e15da6e',
  type: 'Cat',
  name: 'Jazz' 
}
additionalAnimal:  { 
  id: 'e139f1d0-4111-11e7-a34f-2b9b5e15da6e',
  type: 'Cat',
  name: 'Ruby' 
}
whether the selection Ruby has parent:  true

set selection on additionalAnimal successfully.

whether the selection Jazz has parent:  true

set selection on initializedCat successfully.

whether the selection Kitty has parent:  false

/Users/abruzzi/github/mst-issue/node_modules/mobx-state-tree/lib/utils.js:14
        throw new Error("[mobx-state-tree] " + message);
        ^

Error: [mobx-state-tree] Failed to assign a value to a reference; the value should already be part of the same model tree
    at Object.invariant (/Users/abruzzi/github/mst-issue/node_modules/mobx-state-tree/lib/utils.js:14:15)
    at Reference.setNewValue (/Users/abruzzi/github/mst-issue/node_modules/mobx-state-tree/lib/core/reference.js:64:25)
    at Store.Object.defineProperty.set (/Users/abruzzi/github/mst-issue/node_modules/mobx-state-tree/lib/types/property-types/reference-property.js:40:21)
    at Store.setSelection (/Users/abruzzi/github/mst-issue/index.js:33:24)
    at executeAction (/Users/abruzzi/github/mst-issue/node_modules/mobx/lib/mobx.js:663:19)
    at Store.res (/Users/abruzzi/github/mst-issue/node_modules/mobx/lib/mobx.js:654:16)
    at Store.actionInvoker (/Users/abruzzi/github/mst-issue/node_modules/mobx-state-tree/lib/core/action.js:38:27)
    at Store.setSelection (eval at createNamedFunction (/Users/abruzzi/github/mst-issue/node_modules/mobx-state-tree/lib/utils.js:110:12), <anonymous>:3:43)
    at runRawAction (/Users/abruzzi/github/mst-issue/node_modules/mobx-state-tree/lib/core/action.js:5:47)
    at runMiddleWares (/Users/abruzzi/github/mst-issue/node_modules/mobx-state-tree/lib/core/action.js:21:16)
```

**Comment by abruzzihraig** [comment]
@mattiamanzati Ok, I've got it. No need to spend time on checking it. I will create a pull request soon.

**Comment by mattiamanzati** [comment]
@abruzzihraig Thanks! Sorry but so little time these days :( Anyway we know that references are kinda a rogh edge right now, we are working on a way to make them better, but need to change also some mobx internals :)

**Comment by mweststrate** [comment]
@abruzzihraig if you could create a PR with as much additional testcases (preferrably small ones) on (non-working) references as you can imagine, that would be great! Then we can address them all at once hopefully. In the mean time the easiest work around is a property with an identifier, and a manual lookup in a computed property (probably with a setter as well)

**Comment by abruzzihraig** [comment]
@mattiamanzati No worries. Thanks for your great work on MST and MobX. They really help me a lot.

**Comment by abruzzihraig** [comment]
@mweststrate Ok, I will try my best to do that. But I still need you guys to review it later as I don't really understand some details in depth.

**Comment by mweststrate** [comment]
No problem. It will really help us discovering all the ways in which people
want to use this feature, and come up with the best possible solution for
that :)

Op do 25 mei 2017 om 09:40 schreef Yang He <notifications@github.com>:

**Comment by abruzzihraig** [comment]
@mattiamanzati Wait... what the feature do you mean? You mean the `types.reference`? No, I found the bug doesn't come from it. The bug is only about `types.union`. When creating a specific instance from `types.union`, it ignores rest arguments which should be used for getting parent or environment to complex types. Now it basically works as creating primitive types.

Maybe we should change the title of this issue after fixing it. At that time I didn't know which part causes the bug.

**Comment by mattiamanzati** [comment]
Should be fixed in 0.6.3, feel free to reopen if not working :)
